File: lua/pulpero/core/managers/tool/create_file.lua
local Tool = require("tool")
local OSCommands = require('OSCommands')

local function create_file(logger)
    local function execute(params)
        if not params.path then
            return { error = "Path is required" }
        end

        if not params.content then
            return { error = "Content is required" }
        end

        if OSCommands:file_exists(params.path) then
            return { error = "Can not create a file if already exists" }
        end

        logger:debug("Creating file in", { path = params.path })

        local file = io.open(params.path, "w")

        if file ~= nil then
            file:write(params.content)
            file:close()
        else
            return { error = "Can not create the file" }
        end

        return { success = true }
    end

    return Tool.new(
        "create_file",
        "Create a file in the given directory",
        {
            path = {
                type = "string",
                description = "The file path"
            },
            content = {
                type = "string",
                description = "The content of the file"
            }
        },
        "<tool name=\"create_file\" params=\"path=EXACT_PATH, content=CONTENT_OF_THE_FILE\" />",
        execute
    )
end

return {
    create_create_file_tool = create_file,
}

File: core/managers/tool/create_file.py
from core.managers.tool.tool import Tool, ToolResult
import os

from core.util.logger import Logger

def create_file(logger: Logger) -> Tool:
    def execute(params) -> ToolResult:
        if params["path"] == None:
            return ToolResult(False, '',"Path is required")

        if params["content"] == None:
            return ToolResult(False, '',"Content is required")

        if os.path.exists(params["path"]) == True:
            return ToolResult(False, '',"Can not create a file if already exists")

        logger.debug(f"Creating file in path: {params['path']}")

        with open(params["path"], 'w', encoding='utf-8') as f:
            f.write(params["content"])
            f.close()

        return ToolResult(True, 'file created', None)

    return Tool(
        "create_file",
        "Create a file in the given directory",
        {
            "path": {
                "type": "string",
                "title": "path",
                "description": "The file path"
            },
            "content": {
                "type": "string",
                "title": "content",
                "description": "The content of the file"
            }
        },
        ['path', 'content'],
        execute
    )
